vm/devices/storage/scsi_buffers/src/lib.rs

```diff
@@ -204,6 +205,52 @@ impl LockedRange for LockedIoVecs {
     }
 }
 
+/// An implementation of [`MemoryWrite`] that provides semantically
+/// correct results. Specifically, it always returns a `ReadOnly` error
+/// when attempting to write to it.
+pub struct PermissionedMemoryWriter<'a> {
+    range: PagedRange<'a>,
+    writer: PagedRangeWriter<'a>,
+    is_write: bool,
+}
+
+impl PermissionedMemoryWriter<'_> {
+    /// Creates a new memory writer with the given range and guest memory.
+    pub fn new<'a>(
+        range: PagedRange<'a>,
+        guest_memory: &'a GuestMemory,
+        is_write: bool,
+    ) -> PermissionedMemoryWriter<'a> {
+        PermissionedMemoryWriter {
+            range,
+            writer: range.writer(guest_memory),
+            is_write,
+        }
+    }
+}
+
+impl MemoryWrite for PermissionedMemoryWriter<'_> {
+    fn write(&mut self, data: &[u8]) -> Result<(), AccessError> {
+        if self.is_write {
+            self.writer.write(data)
```

Comment:
It would be nice to avoid branches in the hot path here as well. I'd suggest using `PagedRange::empty()` in `new` as suggested elsewhere, and then:
```rust
self.writer.write(data).map_err(|err| {
    if self.is_write { err } else { AccessError::ReadOnly }
})
```